pc/storage/kdtree/kdtree.go

```diff
@@ -116,6 +119,154 @@ func (k *KDTree) searchLeafNode(p mat.Vec3, base []*node) []*node {
+				min = nID
+				continue
+			}
+			if nID == -1 {
+				continue
+			}
+			if k.Vec3At(min)[d] > k.Vec3At(nID)[d] {
+				min = nID
+			}
+		}
+		return min
+	}
+
+	if n.dim == dim {
+		if n.children[0] == nil {
+			return n.id, nil
+		}
+		min, err := k.findMinimumImpl(n.children[0], dim, depth+1)
+		if err != nil {
+			return -1, err
+		}
+		return min, nil
+	}
+
+	min0, err := k.findMinimumImpl(n.children[0], dim, depth+1)
+	if err != nil {
+		return -1, err
+	}
+	min1, err := k.findMinimumImpl(n.children[1], dim, depth+1)
+	if err != nil {
+		return -1, err
+	}
+	return minNode(dim, n.id, min0, min1), nil
+}
+
+func (k *KDTree) stringImpl(n *node, depth int) string {
+	if n != nil {
+		s := k.stringImpl(n.children[1], depth+1)
+		s += fmt.Sprintf(strings.Repeat(" ", 10*depth)+"-> (%d,%d) %v\n", n.id, n.dim, k.Vec3At(n.id))
+		s += k.stringImpl(n.children[0], depth+1)
+		return s
+	}
+	return ""
+}
+
+func (k *KDTree) String() string {
+	return k.stringImpl(k.root, 0)
+}
+
+func (k *KDTree) deleteNodeImpl(n *node, pID int, depth int) (*node, error) {
+	if n == nil {
+		return nil, nil
+	}
+
+	if pID == n.id {
+		if n.children[1] != nil {
+			minNodeID, err := k.findMinimumImpl(n.children[1], n.dim, depth)
+			if err != nil {
+				return nil, err
+			}
+			child, err := k.deleteNodeImpl(n.children[1], minNodeID, depth+1)
+			if err != nil {
+				return nil, err
+			}
+			n.id = minNodeID
+			n.children[1] = child
+		} else if n.children[0] != nil {
+			minNodeID, err := k.findMinimumImpl(n.children[0], n.dim, depth)
+			if err != nil {
+				return nil, err
+			}
+			child, err := k.deleteNodeImpl(n.children[0], minNodeID, depth+1)
+			if err != nil {
+				return nil, err
+			}
+			n.id = minNodeID
+			n.children[0] = nil
+			n.children[1] = child
+		} else {
+			return nil, nil
+		}
+		return n, nil
+	}
+
+	pointAtNode := k.Vec3At(n.id)
+	p := k.Vec3At(pID)
+	if p[n.dim] < pointAtNode[n.dim] {
+		child, err := k.deleteNodeImpl(n.children[0], pID, depth+1)
+		if err != nil {
+			return nil, err
+		}
+		n.children[0] = child
+	} else {
+		child, err := k.deleteNodeImpl(n.children[1], pID, depth+1)
+		if err != nil {
+			return nil, err
+		}
+		n.children[1] = child
+	}
+	return n, nil
+}
+
+func (k *KDTree) DeletePoint(pID int) error {
+	if pID < 0 || pID > k.Len()-1 {
+		return fmt.Errorf("%d does not correspond to any point in the tree", pID)
+	}
+	n, err := k.deleteNodeImpl(k.root, pID, 0)
+	if err != nil {
+		return err
+	}
+	k.root = n
+	return nil
+}
+
+func equal(n1, n2 *node, it1, it2 pc.Vec3RandomAccessor) bool {
+	if n1 == nil && n2 == nil {
+		return true
+	}
+	if n1 == nil && n2 != nil {
+		return false
+	}
+	if n2 == nil && n1 != nil {
+		return false
+	}
+	if n1.id != n2.id || n1.dim != n2.dim || !it1.Vec3At(n1.id).Equal(it2.Vec3At(n2.id)) {
+		return false
+	}
+	return equal(n1.children[0], n2.children[0], it1, it2) &&
+		equal(n1.children[1], n2.children[1], it1, it2)
+}
+
+func (k *KDTree) Equal(k2 *KDTree) bool {
```

Comment:
I'm not very sure `KDTree.Equal` will be used.
If it is not going to be used, I think it's better to be removed to reduce the dead code.